### ISSUE TITLE
about gradient

### ISSUE THREAD

**Issue by rookiez7** [SilmarilBearer/HausdorffLoss#5]
I'm sorry to bother you.
I have a question,you use `torch.no_grad()` in function `distance_field`,The whole process is `pre->pre_dt->loss`, Doesn't `torch.no_grad()` truncate the gradient?Why add this operation can also be trained?Doesn't `pred.cpu().numpy().float()` truncate the gradient?
I'm very confused about this and look forward to your reply and answer

**Comment by Kent0n-Li** [comment]
> 很抱歉打扰你。 我有个问题，你`torch.no_grad()`在函数中使用`distance_field`，整个过程是`pre->pre_dt->loss`，不`torch.no_grad()`截断梯度吗？为什么加这个操作也可以训练？不`pred.cpu().numpy().float()`截断梯度？ 我对此很困惑，期待您的回复和解答

same question

**Comment by Kent0n-Li** [comment]
> I'm sorry to bother you. I have a question,you use `torch.no_grad()` in function `distance_field`,The whole process is `pre->pre_dt->loss`, Doesn't `torch.no_grad()` truncate the gradient?Why add this operation can also be trained?Doesn't `pred.cpu().numpy().float()` truncate the gradient? I'm very confused about this and look forward to your reply and answer

same question

**Comment by rookiez7** [comment]
我目前的理解是，这个`distance_field`当做是一个类似于权重的的东西，但这个值会根据你的预测结果而变化。其次最终进去训练的是mse的变大的loss即pred_erro和这个`diatance_field`的组合。
```
        pred_error = (pred - target) ** 2
        distance = pred_dt ** self.alpha + target_dt ** self.alpha
        dt_field = pred_error * distance
```
这是我询问到的，原作者的解释。你可以参考一下
"p" is usually already binary but "q" should be thresholded.    With regard to differentiation, as you may know the distance transform is highly non-differentiable. We treat it as constant between iterations.

**Comment by Kent0n-Li** [comment]
> 

Thanks!

**Comment by zlyin** [comment]
Also pretty puzzled about this part. Thanks for your information though.
